metals/src/main/scala/scala/meta/internal/builds/SbtBuildTool.scala

```diff
@@ -194,18 +197,30 @@ case class SbtBuildTool(
       shellRunner: ShellRunner,
       workspace: AbsolutePath,
       languageClient: LanguageClient,
+      restartSbtBuildServer: () => Future[Unit],
   )(implicit ex: ExecutionContext): Future[Unit] =
     if (checkCorrectJavaVersion(workspace, userConfig().javaHome)) {
       Future.successful(())
     } else {
-      languageClient
-        .showMessageRequest(Messages.SbtServerJavaHomeUpdate.params())
-        .asScala
-        .flatMap {
-          case Messages.SbtServerJavaHomeUpdate.restart =>
-            shutdownBspServer(shellRunner).ignoreValue
-          case _ => Future.successful(())
-        }
+      val promise = Promise[Unit]()
+      val future: Future[Unit] =
+        languageClient
+          .showMessageRequest(Messages.SbtServerJavaHomeUpdate.params())
+          .asScala
+          .flatMap {
+            case Messages.SbtServerJavaHomeUpdate.restart =>
+              if (promise.isCompleted) restartSbtBuildServer()
```

Comment:
completing the promise will only happen after this future is finished, so this condition will always be false, no?

Reply:
No, I don't think you're right. The promise is completed by this `(...).withTimeout(...)`, so a **new** future with `await` inside. The original Future, won't get cancelled, 'cause you can't cancel futures in Scala, so this is the code that will be executed when the user chooses `restart` after the timeout. (Which you can do in VSCode, 'cause the notifications are stored.) Am I wrong?

Reply:
yes, you're right. I didn't think about the after 15 seconds part. Might be good to add some comments, since it doesn't seem easy to follow.